src/crimson/os/seastore/journal.cc

```diff
@@ -118,35 +119,66 @@ Journal::write_record_ret Journal::write_record(
     target);
 
   auto segment_id = current_journal_segment->get_segment_id();
+  submit_queue.emplace_back(std::make_tuple(target, current_journal_segment, std::move(to_write)));
 
-  // Start write under the current exclusive stage, but wait for it
-  // in the device_submission concurrent stage to permit multiple
-  // overlapping writes.
-  auto write_fut = current_journal_segment->write(target, to_write);
   return handle.enter(write_pipeline->device_submission
-  ).then([write_fut = std::move(write_fut)]() mutable {
-    return std::move(write_fut
-    ).handle_error(
-      write_record_ertr::pass_further{},
-      crimson::ct_error::assert_all{
-	"Invalid error in Journal::write_record"
-      }
-    );
-  }).safe_then([this, &handle] {
+  ).then([this, &handle, target, segment_id] {
     return handle.enter(write_pipeline->finalize);
-  }).safe_then([this, target, segment_id] {
+  }).then([this, target, segment_id]() -> write_record_ret {
     logger().debug(
       "write_record: commit target {}",
       target);
     if (segment_id == current_journal_segment->get_segment_id()) {
       assert(committed_to < target);
       committed_to = target;
     }
-    return write_record_ret(
-      write_record_ertr::ready_future_marker{},
-      paddr_t{
-	segment_id,
-	target});
+    if(!write_queue.empty() || submit_queue.empty()) {
+      return write_record_ret(
+        write_record_ertr::ready_future_marker{},
+        paddr_t{
+          segment_id,
+          target});
+    } else {
+      write_queue.swap(submit_queue);
+      logger().error("write_queue size = {}", write_queue.size());
+      return crimson::repeat([this] {
+        bufferlist bl;
+        auto tuple = write_queue.front();
+        auto [write_target, write_journal_segment, to_write] = tuple;
+        bl.append(to_write);
+        write_queue.pop_front();
+        while (!write_queue.empty()) {
+          auto tmp_tuple = write_queue.front();
+          auto [tmp_target, tmp_journal_segment, tmp_write] = tmp_tuple;
+          if (write_journal_segment == tmp_journal_segment) {
+            bl.append(tmp_write);
+            write_queue.pop_front();
```

Comment:
If the segment changes, I think we want to put the remaining entries back in the queue.  Otherwise, we needlessly expand the latency on this specific write.

Reply:
you mean put them back into the front of submit_queue and write with the remaining bufferlist in submit_queue?  they will be the same segment.

Reply:
I mean that we put the entries for the next segment back in the queue and just do the write to the current segment.